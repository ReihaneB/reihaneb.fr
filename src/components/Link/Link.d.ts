import React from 'react';
import { StaticImport } from 'next/dist/shared/lib/get-img-props';

interface LinkProps {
  to: string;
  color?: 'primary' | 'secondary' | 'transparent';
  size?: 'small' | 'medium';
  children: React.ReactNode;
  rounded?: boolean;
  Icon?: StaticImport | string | null;
  autoFocus?: boolean;
  forwardedRef?: React.RefObject<HTMLButtonElement> | null;
  openInNewTab?: boolean;
  fullWidth?: boolean;
}
