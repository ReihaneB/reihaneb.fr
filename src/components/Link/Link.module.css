.root {
  display: inline-flex;
  align-items: center;
  padding: 12px 16px;
  text-align: center;
  text-decoration: none;
  cursor: pointer;
  border-radius: 4px;
  transition: transform 0.2s cubic-bezier(0, 1.88, 0.82, 1.77);
  user-select: none;
  -webkit-touch-callout: none;
}

.root:focus-visible {
  outline: none;
  box-shadow: 0 0 0 4px var(--focus-blue);
}

.root:active {
  transform: translateY(2px);
}

.rounded {
  border-radius: 9999px;
}

.icon {
  margin-right: 8px;
}

.primary {
  background-color: var(--blue-light);
  color: var(--blue);
}

.primary:hover {
  background-color: var(--blue-lighter);
}

.secondary {
  background-color: var(--blue);
  background-image: linear-gradient(to right, var(--cyan), rgba(255,255,255,0));
  transition: background-color 0.4s;
  color: var(--white);
}

.secondary:hover {
  background-color: var(--cyan);
}

.small {
  padding: 0 8px;
}

