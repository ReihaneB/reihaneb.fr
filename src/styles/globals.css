:root {
  /* colors */
  --white: #f5f5f7;
  --cyan: #06b6d4;
  --blue: #2756df;
  --blue-lighter: #cee5ff;
  --blue-light: #dbeafe;
  --blue-dark: #1e40af;
  --focus-blue: #93c5fd;
  --grey: #6b7280;
  --grey-light: #e5e7eb;

  /* container */
  --container-padding: 16px;
  --container-width: min(calc(100% - var(--container-padding) * 2), 1200px);;

  @media (width <= 1199px) {
    --container-width: min(928px, calc(100% - var(--container-padding) * 2));;
  }

  @media (width <= 768px) {
    --container-width: min(480px, calc(100% - var(--container-padding) * 2));
  }
}

* {
  box-sizing: border-box;
  padding: 0;
  margin: 0;
  border: 0;
  font: inherit;
  vertical-align: baseline;
}

html {
  overflow-x: hidden;
  box-sizing: border-box;
  width: 100%;
  padding-left: calc(50% - var(--container-width) / 2);
  padding-right: calc(50% - var(--container-width) / 2);
}

details,
figcaption,
figure,
footer,
hgroup,
menu,
nav {
  display: block,
}

ol,
ul {
  list-style: none;
}

blockquote,
q {
  quotes: none;
}

blockquote::before,
blockquote::after,
q::before,
q::after {
  content: '';
  content: none;
}

table {
  border-collapse: collapse;
  border-spacing: 0;
}

a {
  color: inherit;
  text-decoration: none;
}

h1 {
  font-size: 48px;
  line-height: 56px;
  font-weight: bold;
}

h2 {
  font-size: 24px;
  line-height: 30px;
  font-weight: bold;
}

h3 {
  font-size: 20px;
  line-height: 24px;
}

h4 {
  font-size: 14px;
  line-height: 18px;
}

h5 {
  font-size: 12px;
  line-height: 16px;
}

p {
  font-size: 16px;
  line-height: 24px;
}

strong {
  font-weight: bold;
}

code {
  font-family: 'Courier New', Courier, monospace;
  background-color: var(--grey-light);
  padding: 2px 4px;
  border-radius: 4px;
}