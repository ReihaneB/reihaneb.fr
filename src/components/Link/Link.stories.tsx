import type { Meta, StoryObj } from '@storybook/react';
import { expect, userEvent, waitFor, within } from '@storybook/test';

import { lokiPlayDecorator } from '@/testUtils/loki/play-decorator';
import IconArrow from '@/styles/icons/arrows/arrow.svg';

import Link from './Link';

import styles from './Link.stories.module.css';

const meta: Meta<typeof Link> = {
  title: 'Components/Link',
  component: Link,
};

export default meta;
type Story = StoryObj<typeof Link>;

const defaultArgs = {
  to: '/',
  children: 'Click on the link',
  autoFocus: false,
  forwardedRef: null,
  fullWidth: false,
};

export const Playground: Story = {
  args: {
    ...defaultArgs,
  },
};

export const Primary: Story = {
  args: {
    ...defaultArgs,
  },
};

export const PrimaryHover: Story = {
  name: 'Primary - Hover',
  args: {
    ...defaultArgs,
  },
  parameters: {
    pseudo: { hover: true },
  },
};

export const Secondary: Story = {
  args: {
    ...defaultArgs,
    color: 'secondary',
  },
};

export const SecondaryHover: Story = {
  name: 'Secondary - Hover',
  args: {
    ...defaultArgs,
    color: 'secondary',
  },
  parameters: {
    pseudo: { hover: true },
  },
};

export const Active: Story = {
  args: {
    ...defaultArgs,
  },
  parameters: {
    pseudo: { active: true },
  },
};

export const Focus: Story = {
  args: {
    ...defaultArgs,
  },
  parameters: {
    pseudo: { focusVisible: true },
  },
};

export const PrimaryRounded: Story = {
  args: {
    ...defaultArgs,
    rounded: true,
  },
};

export const SecondaryRounded: Story = {
  args: {
    ...defaultArgs,
    color: 'secondary',
    rounded: true,
  },
};

export const PrimarySmall: Story = {
  args: {
    ...defaultArgs,
    size: 'small',
  },
};

export const PrimarySmallRounded: Story = {
  args: {
    ...defaultArgs,
    size: 'small',
    rounded: true,
  },
};

export const SecondarySmall: Story = {
  args: {
    ...defaultArgs,
    color: 'secondary',
    size: 'small',
  },
};

export const SecondarySmallRounded: Story = {
  args: {
    ...defaultArgs,
    color: 'secondary',
    size: 'small',
    rounded: true,
  },
};

export const FullWidth: Story = {
  args: {
    ...defaultArgs,
    fullWidth: true,
  },
};

export const WithIcon: Story = {
  args: {
    ...defaultArgs,
    Icon: IconArrow,
  },
};

export const SmallWithIcon: Story = {
  args: {
    ...defaultArgs,
    size: 'small',
    Icon: IconArrow,
  },
};

export const RoundedWithIcon: Story = {
  args: {
    ...defaultArgs,
    rounded: true,
    Icon: IconArrow,
  },
};

export const SmallRoundedWithIcon: Story = {
  args: {
    ...defaultArgs,
    size: 'small',
    rounded: true,
    Icon: IconArrow,
  },
};

export const AutoFocus: Story = {
  args: {
    ...defaultArgs,
    autoFocus: true,
  },
  play: lokiPlayDecorator(
    async ({ canvasElement }) => {
      const canvas = within(canvasElement);

      await waitFor(() => expect(canvas.getByTestId('link-component')).toHaveFocus());
    }
  ),
};

export const ClickMultipleTimes: Story = {
  args: {
    ...defaultArgs,
  },
  play: lokiPlayDecorator(
    async ({ canvasElement, step }) => {
      const canvas = within(canvasElement);

      await step('Click the link 2 times', async () => {
        await userEvent.click(canvas.getByTestId('link-component'));
        await userEvent.click(canvas.getByTestId('link-component'));
      });

      // TODO: Except Actions when it will be implemented with Storybook Interaction
    }
  ),
};

export const PressEnter: Story = {
  args: {
    ...defaultArgs,
  },
  play: lokiPlayDecorator(
    async ({ canvasElement, step }) => {
      const canvas = within(canvasElement);

      await step('Press the link 2 times', async () => {
        await userEvent.type(canvas.getByTestId('link-component'), '{Enter}');
        await userEvent.type(canvas.getByTestId('link-component'), '{Enter}');
      });

      // TODO: Except Actions when it will be implemented with Storybook Interaction
    }
  ),
};

export const FocusOnClick: Story = {
  name: 'Focused when clicked',
  args: {
    ...defaultArgs,
  },
  play: lokiPlayDecorator(
    async ({ canvasElement }) => {
      const canvas = within(canvasElement);

      await userEvent.click(canvas.getByTestId('link-component'));

      await waitFor(() => expect(canvas.getByTestId('link-component')).toHaveFocus());
    }
  ),
};

export const FocusOnTab: Story = {
  name: 'Focused when tabbed',
  args: {
    ...defaultArgs,
  },
  play: lokiPlayDecorator(
    async ({ canvasElement }) => {
      const canvas = within(canvasElement);

      await userEvent.tab();

      await waitFor(() => expect(canvas.getByTestId('link-component')).toHaveFocus());
    }
  ),
};

export function LongText() {
  return (
    <div className={styles.container}>
      <Link
        to="/"
      >
        This is a very long text that should be going to the next line
      </Link>
    </div>
  );
}

LongText.storyName = 'Long text';

export function LongTextWithIcon() {
  return (
    <div className={styles.container}>
      <Link
        to="/"
        Icon={IconArrow}
      >
        This is a very long text that should be going to the next line
      </Link>
    </div>
  );
}

LongTextWithIcon.storyName = 'Long text with icon';
