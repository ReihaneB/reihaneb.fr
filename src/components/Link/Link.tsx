'use client';

import { memo } from 'react';
import NextLink from 'next/link';
import Image from 'next/image';
import classNames from 'classnames';

import { useFocusable } from '@/hooks/useFocusable/useFocusable';

import styles from './Link.module.css';
import type { LinkProps } from './Link.d';

function Link({
  to,
  color = 'primary',
  size = 'medium',
  rounded = false,
  children,
  Icon = null,
  autoFocus = false,
  forwardedRef = null,
  openInNewTab = false,
  fullWidth = false,
  ...rest
}: LinkProps) {
  const { ref: elementRef } = useFocusable(forwardedRef, { autoFocus });

  return (
    <NextLink
      href={to}
      className={
        classNames(styles.root, {
          [styles.primary]: color === 'primary',
          [styles.secondary]: color === 'secondary',
          [styles.small]: size === 'small',
          [styles.rounded]: rounded,
          [styles.fullWidth]: fullWidth,
        })
      }
      ref={elementRef}
      data-testid="link-component"
      target={openInNewTab ? '_blank' : '_self'}
      rel={openInNewTab ? 'noreferrer' : undefined}
      {...rest}
    >
      {Icon && (
        <Image
          src={Icon}
          alt={`IcÃ´ne de lien vers ${to}`}
          width={12}
          height={12}
          className={styles.icon}
        />
      )}
      {children}
    </NextLink>
  );
}

export default memo(Link);
