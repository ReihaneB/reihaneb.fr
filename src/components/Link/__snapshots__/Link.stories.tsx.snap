// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/Link Active smoke-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <a class="src-components-Link-Link-module__root--ggyqZ src-components-Link-Link-module__primary--hzP7j"
     data-testid="link-component"
     target="_self"
     href="/"
  >
    Click on the link
  </a>
</main>
`;

exports[`Components/Link AutoFocus play-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <a class="src-components-Link-Link-module__root--ggyqZ src-components-Link-Link-module__primary--hzP7j"
     data-testid="link-component"
     target="_self"
     href="/"
  >
    Click on the link
  </a>
</main>
`;

exports[`Components/Link ClickMultipleTimes play-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <a class="src-components-Link-Link-module__root--ggyqZ src-components-Link-Link-module__primary--hzP7j"
     data-testid="link-component"
     target="_self"
     href="/"
  >
    Click on the link
  </a>
</main>
`;

exports[`Components/Link Focus smoke-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <a class="src-components-Link-Link-module__root--ggyqZ src-components-Link-Link-module__primary--hzP7j"
     data-testid="link-component"
     target="_self"
     href="/"
  >
    Click on the link
  </a>
</main>
`;

exports[`Components/Link FocusOnClick play-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <a class="src-components-Link-Link-module__root--ggyqZ src-components-Link-Link-module__primary--hzP7j"
     data-testid="link-component"
     target="_self"
     href="/"
  >
    Click on the link
  </a>
</main>
`;

exports[`Components/Link FocusOnTab play-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <a class="src-components-Link-Link-module__root--ggyqZ src-components-Link-Link-module__primary--hzP7j"
     data-testid="link-component"
     target="_self"
     href="/"
  >
    Click on the link
  </a>
</main>
`;

exports[`Components/Link LongText smoke-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <div class="src-components-Link-Link-stories-module__container--Kpt8t">
    <a class="src-components-Link-Link-module__root--ggyqZ src-components-Link-Link-module__primary--hzP7j"
       data-testid="link-component"
       target="_self"
       href="/"
    >
      This is a very long text that should be going to the next line
    </a>
  </div>
</main>
`;

exports[`Components/Link LongTextWithIcon smoke-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <div class="src-components-Link-Link-stories-module__container--Kpt8t">
    <a class="src-components-Link-Link-module__root--ggyqZ src-components-Link-Link-module__primary--hzP7j"
       data-testid="link-component"
       target="_self"
       href="/"
    >
      <img alt="Icône de lien vers /"
           loading="lazy"
           width="12"
           height="12"
           decoding="async"
           data-nimg="1"
           class="src-components-Link-Link-module__icon--B0cve"
           srcset="/static/media/arrow.55ce8cbb.svg?w=16&amp;q=75 1x, /static/media/arrow.55ce8cbb.svg?w=32&amp;q=75 2x"
           src="/static/media/arrow.55ce8cbb.svg?w=32&amp;q=75"
           style="color: transparent;"
      >
      This is a very long text that should be going to the next line
    </a>
  </div>
</main>
`;

exports[`Components/Link Playground smoke-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <a class="src-components-Link-Link-module__root--ggyqZ src-components-Link-Link-module__primary--hzP7j"
     data-testid="link-component"
     target="_self"
     href="/"
  >
    Click on the link
  </a>
</main>
`;

exports[`Components/Link PressEnter play-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <a class="src-components-Link-Link-module__root--ggyqZ src-components-Link-Link-module__primary--hzP7j"
     data-testid="link-component"
     target="_self"
     href="/"
  >
    Click on the link
  </a>
</main>
`;

exports[`Components/Link Primary smoke-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <a class="src-components-Link-Link-module__root--ggyqZ src-components-Link-Link-module__primary--hzP7j"
     data-testid="link-component"
     target="_self"
     href="/"
  >
    Click on the link
  </a>
</main>
`;

exports[`Components/Link PrimaryHover smoke-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <a class="src-components-Link-Link-module__root--ggyqZ src-components-Link-Link-module__primary--hzP7j"
     data-testid="link-component"
     target="_self"
     href="/"
  >
    Click on the link
  </a>
</main>
`;

exports[`Components/Link PrimaryRounded smoke-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <a class="src-components-Link-Link-module__root--ggyqZ src-components-Link-Link-module__primary--hzP7j src-components-Link-Link-module__rounded--zQTlH"
     data-testid="link-component"
     target="_self"
     href="/"
  >
    Click on the link
  </a>
</main>
`;

exports[`Components/Link PrimarySmall smoke-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <a class="src-components-Link-Link-module__root--ggyqZ src-components-Link-Link-module__primary--hzP7j src-components-Link-Link-module__small--nBuK2"
     data-testid="link-component"
     target="_self"
     href="/"
  >
    Click on the link
  </a>
</main>
`;

exports[`Components/Link PrimarySmallRounded smoke-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <a class="src-components-Link-Link-module__root--ggyqZ src-components-Link-Link-module__primary--hzP7j src-components-Link-Link-module__small--nBuK2 src-components-Link-Link-module__rounded--zQTlH"
     data-testid="link-component"
     target="_self"
     href="/"
  >
    Click on the link
  </a>
</main>
`;

exports[`Components/Link RoundedWithIcon smoke-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <a class="src-components-Link-Link-module__root--ggyqZ src-components-Link-Link-module__primary--hzP7j src-components-Link-Link-module__rounded--zQTlH"
     data-testid="link-component"
     target="_self"
     href="/"
  >
    <img alt="Icône de lien vers /"
         loading="lazy"
         width="12"
         height="12"
         decoding="async"
         data-nimg="1"
         class="src-components-Link-Link-module__icon--B0cve"
         srcset="/static/media/arrow.55ce8cbb.svg?w=16&amp;q=75 1x, /static/media/arrow.55ce8cbb.svg?w=32&amp;q=75 2x"
         src="/static/media/arrow.55ce8cbb.svg?w=32&amp;q=75"
         style="color: transparent;"
    >
    Click on the link
  </a>
</main>
`;

exports[`Components/Link Secondary smoke-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <a class="src-components-Link-Link-module__root--ggyqZ src-components-Link-Link-module__secondary--tiYlA"
     data-testid="link-component"
     target="_self"
     href="/"
  >
    Click on the link
  </a>
</main>
`;

exports[`Components/Link SecondaryHover smoke-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <a class="src-components-Link-Link-module__root--ggyqZ src-components-Link-Link-module__secondary--tiYlA"
     data-testid="link-component"
     target="_self"
     href="/"
  >
    Click on the link
  </a>
</main>
`;

exports[`Components/Link SecondaryRounded smoke-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <a class="src-components-Link-Link-module__root--ggyqZ src-components-Link-Link-module__secondary--tiYlA src-components-Link-Link-module__rounded--zQTlH"
     data-testid="link-component"
     target="_self"
     href="/"
  >
    Click on the link
  </a>
</main>
`;

exports[`Components/Link SecondarySmall smoke-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <a class="src-components-Link-Link-module__root--ggyqZ src-components-Link-Link-module__secondary--tiYlA src-components-Link-Link-module__small--nBuK2"
     data-testid="link-component"
     target="_self"
     href="/"
  >
    Click on the link
  </a>
</main>
`;

exports[`Components/Link SecondarySmallRounded smoke-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <a class="src-components-Link-Link-module__root--ggyqZ src-components-Link-Link-module__secondary--tiYlA src-components-Link-Link-module__small--nBuK2 src-components-Link-Link-module__rounded--zQTlH"
     data-testid="link-component"
     target="_self"
     href="/"
  >
    Click on the link
  </a>
</main>
`;

exports[`Components/Link SmallRoundedWithIcon smoke-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <a class="src-components-Link-Link-module__root--ggyqZ src-components-Link-Link-module__primary--hzP7j src-components-Link-Link-module__small--nBuK2 src-components-Link-Link-module__rounded--zQTlH"
     data-testid="link-component"
     target="_self"
     href="/"
  >
    <img alt="Icône de lien vers /"
         loading="lazy"
         width="12"
         height="12"
         decoding="async"
         data-nimg="1"
         class="src-components-Link-Link-module__icon--B0cve"
         srcset="/static/media/arrow.55ce8cbb.svg?w=16&amp;q=75 1x, /static/media/arrow.55ce8cbb.svg?w=32&amp;q=75 2x"
         src="/static/media/arrow.55ce8cbb.svg?w=32&amp;q=75"
         style="color: transparent;"
    >
    Click on the link
  </a>
</main>
`;

exports[`Components/Link SmallWithIcon smoke-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <a class="src-components-Link-Link-module__root--ggyqZ src-components-Link-Link-module__primary--hzP7j src-components-Link-Link-module__small--nBuK2"
     data-testid="link-component"
     target="_self"
     href="/"
  >
    <img alt="Icône de lien vers /"
         loading="lazy"
         width="12"
         height="12"
         decoding="async"
         data-nimg="1"
         class="src-components-Link-Link-module__icon--B0cve"
         srcset="/static/media/arrow.55ce8cbb.svg?w=16&amp;q=75 1x, /static/media/arrow.55ce8cbb.svg?w=32&amp;q=75 2x"
         src="/static/media/arrow.55ce8cbb.svg?w=32&amp;q=75"
         style="color: transparent;"
    >
    Click on the link
  </a>
</main>
`;

exports[`Components/Link WithIcon smoke-test 1`] = `
<main class="-storybook-preview-module__root--fQckk inter-normal">
  <a class="src-components-Link-Link-module__root--ggyqZ src-components-Link-Link-module__primary--hzP7j"
     data-testid="link-component"
     target="_self"
     href="/"
  >
    <img alt="Icône de lien vers /"
         loading="lazy"
         width="12"
         height="12"
         decoding="async"
         data-nimg="1"
         class="src-components-Link-Link-module__icon--B0cve"
         srcset="/static/media/arrow.55ce8cbb.svg?w=16&amp;q=75 1x, /static/media/arrow.55ce8cbb.svg?w=32&amp;q=75 2x"
         src="/static/media/arrow.55ce8cbb.svg?w=32&amp;q=75"
         style="color: transparent;"
    >
    Click on the link
  </a>
</main>
`;
